### DISCUSSION TITLE
Support for strict ISO8601 formats

### DISCUSSION THREAD

**Discussion by smarie** [pydantic/pydantic#2951] [Feature Request]
ISO8601 supports many formats: year, month, week, week with week day, date without year, date, datetime, time only. 

Several of these formats may also have timezone (utc) or time offset (+0100, etc.) information appended.

We could probably propose "strict iso8601" objects in pydantic, similar to StrictStr, StrictFloat etc. The most common formats could be provided "out of the box", while other formats could be created by inheriting from a parent class and overriding some class fields, as in the ConstrainedString class in pydantic

Could this be of interest to pydantic ? Or should it rather be part of a plugin ?

**Comment by antonagestam** [comment]
@smarie As an alternative that should work with Pydantic out of the box, at least some of those should be trivial to implement with [phantom-types](https://github.com/antonagestam/phantom-types).

For instance, an ISO8601 year could be expressed as:

```python
import phantom.re
import re

class Year(str, phantom.re.FullMatch, pattern=re.compile(r"[0-9]{4}")):
    ...
```

I'd be open to the idea of adding an `iso8601` module to phantom-types.

**Reply by smarie** [reply]
Thanks for the proposal @antonagestam ! I did not know phantom-types, it looks interesting. 
However my proposal is still focused on pydantic because this is how the JSON Schema for fastapi is generated.

**Reply by antonagestam** [reply]
Gotcha, just so you know though, phantom-types supports the JSON schema facilities of Pydantic 🙂

**Reply by smarie** [reply]
Good to know ! You'l have to convince FastAPI now :) so that they officially support it. 

By the way phantom-types reminds me of [vtypes](https://smarie.github.io/python-vtypes/) - long time ago already, but just for fun

**Reply by antonagestam** [reply]
@smarie Unless I'm missing something there should already be support all the way as both FastAPI and phantom-types support pydantic.

I tried with this example:

```python
import re
from pydantic import BaseModel
from fastapi import FastAPI
import phantom.re
from phantom.ext.iso3166 import CountryCode


class Year(phantom.re.FullMatch, pattern=re.compile(r"[0-9]{4}")):
    @classmethod
    def __schema__(cls):
        return {
            **super().__schema__(),
            "format": "An ISO8601 year, e.g. in the format NNNN",
        }


class User(BaseModel):
    name: str
    country: CountryCode
    birth_year: Year


app = FastAPI()

@app.post("/")
def create_user(user: User):
    ...
```

And it yields a redoc schema like this:

<img width="1514" alt="Skärmavbild 2021-07-09 kl  11 27 29" src="https://user-images.githubusercontent.com/218551/125056778-dd7cc000-e0a8-11eb-90b4-bb1d2e6b053b.png">


> By the way phantom-types reminds me of vtypes - long time ago already, but just for fun

Oh, very cool! I will definitely look through that for inspiration 👍

**Reply by smarie** [reply]
Nice ! Thanks for this example
